pdd/PDD-multistream.md

```diff
@@ -0,0 +1,139 @@
+# PDD for multistream (example/use case)
+
+## Protocol
+
+The motivation and base of the protocol can be found here:
+
+- https://github.com/ipfs/specs/blob/wire/protocol/network/wire.md#multistream---self-describing-protocol-stream
+- https://github.com/ipfs/specs/blob/wire/protocol/network/wire.md#multistream-selector---self-describing-protocol-stream-selector
+
+The multistream protocol does not cover:
+
+- discovering participants, selecting transports, and establishing connections
+- managing the state of the connection
+
+multistream enables several types of streams to be used over one single stream, like an intelligent message broker that offers the ability to negotiate the protocol and version that is going to be used. To simplify, a visual representation can be:
+
+```
+┌ ─ ─ ─ ─ ─ ─ ┌ ─ ─ ─ ─ ─ ─ ┐┌ ─ ┐
+ dht-id/1.0.1│ bitswap/1.2.3  ...
+└ ─ ─ ─ ─ ─ ─ └ ─ ─ ─ ─ ─ ─ ┘└ ─ ┘
+┌────────────────────────────────┐
+│ multistream-select             │
+└────────────────────────────────┘
+┌────────────────────────────────┐
+│transport                       │
+│  [TCP, UDP, uTP, ...]          │
+└────────────────────────────────┘
+```
+
+multistream doesn't cover stream multiplexing over the same connection, however, we can achieve this by leveraging functionality offered by SPDY or HTTP/2.
```

Comment:
On Sat, Jun 13, 2015 at 08:43:29PM -0700, Juan Batiz-Benet wrote:
> > Without a stream-multiplexing transport layer, using
> > `multistream-select` will be difficult (impossible?).
> 
> no. it will not. easy, as the examples suggest stream will just be
> locked into that protocol until the stream terminates.

Ah, I was just thinking that multistream-select assumed it would
always be running for stream maintenance.  But looking over [1,2], it
looks like ‘multistream’ is just “write a protocol-describing header
line before sending any other data down the stream”.  So that's fine.
Then ‘multistream-select’ adds the ‘ls’ listing and ‘na\n’.  The
handoff between the two is still a bit murky to me.  If the
uniplex-transport conversation is:

← /multistream/1.0.0
← /bird/3.2.1
← hey, how is it going?

Switching to a /bird/ protocol is a multistream-select situation.  So
it should probalby be:

← /multistream-select/1.0.0
← /bird/3.2.1
← hey, how is it going?

And the listener handlers will be:

1. Open with a multistream-select driver on the socket.
2. Receive ‘/multistream-select/1.0.0’.  Good, that's the language we
   speak.
3. Receive ‘/bird/3.2.1’, that's a protocol we support.  Pop ourselves
   off the socket (and into a protocol stack?), and attach the
   bird-3.2.1 driver.
4. Replay the ‘/bird/3.2.1’ message into the bird driver?
5. Receive ‘hey, how is it going?’ and process with the current bird
   driver.

The sticky bit is (4), which I don't think we actually need.  But if
the ‘/bird/3.2.1’ is only sent on the multistream-select “stream”,
then the bird stream itself is just ‘hey, how is it going?’, which
isn't a multistream protocol.  That's fine, it doesn't have to be a
multistream protocol, but if the only multistream protocol is
multistream-select, it seems a bit odd ;).

I'm also not sure how, in the uniplexed transport, a subprotocol is
supposed to open a new child protocol.  For example, if the logicial
stream spawning structure was like:

← /multistream-select/1.0.0
  ← /bird/3.2.1
    ← /parrot/1.0.0

If you have a parallel stream that's still in multistream-select, you
can just ask for a new stream with the child protocol there:

← /multistream-select/1.0.0
  ← /bird/3.2.1
    … bird stream …
  ← /parrot/1.0.0
    … new parrot stream …

That means the agent requesting the opening needs to reach back up to
the multistream-select stream, but that's ok.  Over uniplexed
transport, it seems like you'd either have to build subprotocol
spawning into your spawning protocol:

← /multistream-select/1.0.0
← /bird/3.2.1
… bird stream …
← /parrot/1.0.0  # handled by bird driver
… new parrot stream …

Or close the parent:

← /multistream-select/1.0.0
← /bird/3.2.1
… bird stream …
← exit  # or whatever the bird-driver needs to shut down
← /parrot/1.0.0  # handled by multistream-select driver
… new parrot stream …

And I'm a bit nervous about both of those choices.  Maybe that's just
stretching multistream-select over unplexed transport too far?

[1]: https://github.com/jbenet/multistream/blob/master/README.md
[2]: https://github.com/jbenet/multistream/tree/master/multistream-select